Produce consume tutorial in scala (#459)
confluentinc/kafka-tutorials: _includes/tutorials/produce-consume-lang/scala/markup/dev/5_0-produce-narrative.adoc

The following Scala class will be the entry point of the producer application.
At `src/main/scala/io/confluent/developer/Producer.scala` add:

+++++
<pre class="snippet"><code class="java">{%
    include_raw tutorials/produce-consume-lang/scala/code/src/main/scala/io/confluent/developer/Producer.scala
%}</code></pre>
+++++

Lets's describe the keys pieces of this program.

- Reminder: we use an externalised config and this app loads its config from the _producer_ bloc
of the `application.conf`

A `Producer#produce` function covers most of the record production.

- `Producer#produce` takes a
`https://kafka.apache.org/25/javadoc/index.html?org/apache/kafka/clients/producer/KafkaProducer.html[KafkaProducer]`,
a topic name, and an instance of book to send into a Kafka topic.

- `Producer#produce` wraps our book with a `ProducerRecord[K, V]`, that's where attache the topic name to the book.

- `Producer#produce` call the `KafkaProducer#send` method and gets back Java future tight to the broker acknowledgement.

- `Producer#produce` use a
`https://kafka.apache.org/25/javadoc/index.html?org/apache/kafka/clients/producer/KafkaProducer.html[CallBack]`

> to confirm the future completion and the record write.

Reviewer: Great details on the code. We've recently added some additional asciidoctor support
.
Could you put the `Producer.produce`  method in a code block with callouts? 

You probably don't need this but [here's an example](https://github.com/confluentinc/kafka-tutorials/blob/master/_includes/tutorials/serialization/kstreams/markup/dev/make-topology.adoc)

> The following Scala class will be the entry point of the producer application.
> At `src/main/scala/io/confluent/developer/Producer.scala` add:

Reviewer 2: ```suggestion
The following Scala class will be the entry point of the producer application.

Add the following class in the file `src/main/scala/io/confluent/developer/Producer.scala`
```

+++++
<pre class="snippet"><code class="java">{%
    include_raw tutorials/produce-consume-lang/scala/code/src/main/scala/io/confluent/developer/Producer.scala
%}</code></pre>
+++++

> Lets's describe the keys pieces of this program.

Reviewer 2: ```suggestion
Lets's describe the key sections of the producer code.
```

> NOTE: we use an externalised config and this app loads its config from the _producer_ bloc
> of the `application.conf`

Reviewer 2: ```suggestion
NOTE: The producer application loads its configuration from the _producer_ block
of the `application.conf`
```

A `Producer#produce` function covers most of the record production.

[source,scala]
----
def produce(producer: KafkaProducer[Bytes, Book], topic: String, book: Book): Future[RecordMetadata] = { //<1>
    val record: ProducerRecord[Bytes, Book] = new ProducerRecord(topic, book) //<2>

    producer.send(record, new Callback { //<3>
      override def onCompletion(metadata: RecordMetadata, exception: Exception): Unit = Option(exception) //<4>
        .map(ex => logger error s"fail to produce record due to: ${ex.getMessage}")
        .getOrElse(logger info s"successfully produced - ${printMetaData(metadata)}")
    })
}
----

> <1> `Producer#produce` takes a
> `https://kafka.apache.org/25/javadoc/index.html?org/apache/kafka/clients/producer/KafkaProducer.html[KafkaProducer]`,
> a topic name, and an instance of book to send into a Kafka topic.

Reviewer 2: ```suggestion
<1> The function accepts a
`https://kafka.apache.org/25/javadoc/index.html?org/apache/kafka/clients/producer/KafkaProducer.html[KafkaProducer]`,
a topic name, and an instance of book to send into a Kafka topic.
```

> <2> `Producer#produce` wraps our book with a `ProducerRecord[K, V]`, that's where attache the topic name to the book.

Reviewer 2: ```suggestion
<2> `Producer#produce` wraps our book with a `ProducerRecord[K, V]`, that's where attach the topic name to the book.
```

> <3> `Producer#produce` call the `KafkaProducer#send` method and gets back Java future tight to the broker
> acknowledgement.

Reviewer 2: ```suggestion
<3> The `KafkaProducer#send` method  is called on the `producer` instance, returning a Java future which will contain the broker response.
```

> <4> `Producer#produce` use a
> `https://kafka.apache.org/25/javadoc/index.html?org/apache/kafka/clients/producer/KafkaProducer.html[CallBack]`
> to confirm the future completion and the record write.

Reviewer 2: ```suggestion
<4> `Producer#produce` uses a
`https://kafka.apache.org/25/javadoc/index.html?org/apache/kafka/clients/producer/KafkaProducer.html[CallBack]`
to respond to success or failure for the record production.
```

> 🤔 🤔 Now what do we need to call our new `Producer#produce` function?
> 
> - The `KafkaProducer` has types parameter corresponding to its key and value types. Its constructor takes
> serializer / deserializers with the same types.
> 
> - `reflectionSerializer4S[T]` is introduced to lighten the tutorial and quickly get a `Serializer[Book]` (See dev part n°3).
> 
> - We carefully configure, the serializer with a map containing the schema-registry url.
> 
> - The keys will always be null in our example, so we choose an arbitrary type such `Bytes`
> from _kafka.common.utils_ package.
> 
> - Then we instantiate the `KafkaProducer` by passing a Java Properties from the _producer.client-config_ block of
> `application.conf`

Reviewer 2: ```suggestion
In order to utilize the `Producer#produce` function, we construct an instance of our `KafkaProducer` to pass to it.

- The `KafkaProducer` has types parameter corresponding to its key and value types. Its constructor takes
serializer / deserializers with the same types.

- As described above, we utilze the `reflectionSerializer4S[T]` for our `Serializer[Book]` which is configured to connect to the Schema Registry.

- Our Kafka records do not have keys and will always be null, so we choose an arbitrary type such `Bytes`
from the _kafka.common.utils_ package.

- We instantiate the `KafkaProducer` by passing a Java Properties from the _producer.client-config_ block of
`application.conf`
```

That's it! we are ready to produce records.

> NOTE: At the end of the programme we use `KafkaProducer#flush` to check

Reviewer 2: ```suggestion
NOTE: At the end of the program we use `KafkaProducer#flush` to check
```

Author: Sorry 🙋‍♂️🇫🇷🇫🇷🇫🇷🥖🧀
